coding/using_boost.rst

```diff
@@ -4,24 +4,38 @@ Using Boost
 
 .. _cpp_using_boost:
 
-Certain Boost libraries are recommended; they should be used whenever applicable in preference to any other method of accomplishing the same effect. In general, any library that is tagged with *"Standard .... {something}"* in the `Boost library listing <http://www.boost.org/doc/libs>`_ falls into this category, unless the feature is already supported by the currently-mandated LSST ``g++`` version.
-Among others, this category includes:
+A Boost library shall be used in in preference to any other method of accomplishing the same effect only if:
 
-* ``current_function``
-* ``format``
-* ``regex`` (until we adopt ``gcc`` 4.9 as our minimum compiler version, at which point use ``std::regex``)
-* ``test``
+1. the effect cannot be accomplished with a C++11 standard language feature, and
+2. a C++ standard library equivalent is not available, or
+3. the C++ standard library equivalent is not usable with our minimum supported compiler version (i.e. ``gcc`` version 4.8).
 
-Libraries that were accepted before we switched to C++11 (in particular made ``gcc`` 4.8 our minimum supported compiler), but have been or are being replaced by ``std::`` equivalents include:
+In particular the following libraries were accepted before we adopted ``gcc`` 4.8 as our minimum compiler version, but have been or are being replaced by ``std::`` equivalents,
+and are thus no longer accepted, include:
 
 * ``array``: use ``std::array`` from ``<array>``
+* ``bind``: use ``std::bind``
 * ``cstdint``: use ``<cstdint>``
 * ``filesystem``: use ``<filesystem>``
+* ``lambda``: use C++11 lambda functions
+* ``lexical_cast``: use ``std::to_string``, ``std::sto``, ``std::stod`` etc.
+* ``math``: use ``<cmath>`` wherever possible
 * ``noncopyable``: use ``= delete`` on the copy constructor and assignment operator
+* ``random``: use ``<random>``
+* ``ref``: use ``std::forward`` and universal / forwarding references
 * ``smart_ptr``: use ``std::shared_ptr`` and ``std::unique_ptr`` from ``<memory>`` instead of ``boost::shared_ptr`` and ``boost::scoped_ptr``
-* ``static_assert``: use ``static_assert``
+* ``static_assert``: use C++11 ``static_assert``
+* ``tuple``: use ``<tuple>``
 * ``type_traits``: use ``<type_traits>``
-* ``unordered_map``: use ``std::unordered_map`` from ``<unordered_map>``
+* ``unordered_map``: use ``<unordered_map>``
+
+Certain Boost libraries are recommended; they should be used whenever applicable in preference to any other method of accomplishing the same effect. In general, any library that is tagged with *"Standard .... {something}"* in the `Boost library listing <http://www.boost.org/doc/libs>`_ falls into this category (unless its use conflicts with the above rules on the availability of standard library / language equivalents).
+Among others, this category includes:
+
+* ``current_function``
+* ``format``
+* ``regex`` (until we adopt ``gcc`` 4.9 as our minimum compiler version, at which point use ``std::regex``)
```

Comment:
Some time back I really tried to give std::regex a go.  Beyond the gcc 
4.8 bustage, I also found it to be missing some very useful things that 
I had liked with boost::regex:

  * named sub-expressions
  * look-behind
  * comments
  * PCRE compatibility in general (std::regex is basically ECMA)

Disappointing :-(

On 06/01/2016 12:16 PM, ktlim wrote:
>
> In coding/using_boost.rst 
> <https://github.com/lsst-dm/dm_dev_guide/pull/39#discussion_r65424603>:
>
> >  * ``type_traits``: use ``<type_traits>``
> > -* ``unordered_map``: use ``std::unordered_map`` from ``<unordered_map>``
> > +* ``unordered_map``: use ``<unordered_map>``
> > +
> > +Certain Boost libraries are recommended; they should be used whenever applicable in preference to any other method of accomplishing the same effect. In general, any library that is tagged with *"Standard .... {something}"* in the `Boost library listing <http://www.boost.org/doc/libs>`_ falls into this category (unless its use conflicts with the above rules on the availability of standard library / language equivalents).
> > +Among others, this category includes:
> > +
> > +* ``current_function``
> > +* ``format``
> > +* ``regex`` (until we adopt ``gcc`` 4.9 as our minimum compiler version, at which point use ``std::regex``)
>
> I thought @fritzm <https://github.com/fritzm> still had potential 
> problems with |std::regex| due to lack of features.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/lsst-dm/dm_dev_guide/pull/39/files/1c638713c2dda7ef7aa35e191d32a28f07b846da#r65424603>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe/ABdbDn4uij_ux3rWZ0IzHv1LMlDz99TIks5qHdqKgaJpZM4Ir4Md>.
>